tracing-opentelemetry/src/main/java/io/micronaut/tracing/opentelemetry/instrument/util/TracingPublisher.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.tracing.opentelemetry.instrument.util;
+
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.async.publisher.Publishers;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.util.ClassAndMethod;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+/**
+ * A reactive streams publisher that traces.
+ *
+ * @param <T> the type of element signaled
+ * @author Nemanja Mikic
+ */
+@SuppressWarnings("PublisherImplementation")
+public class TracingPublisher<T> implements Publishers.MicronautPublisher<T> {
+
+    private final Publisher<T> publisher;
+    private final Instrumenter<ClassAndMethod, Object> instrumenter;
+    @Nullable
+    private final ClassAndMethod classAndMethod;
+    private final Context parentContext;
+
+    /**
+     * @param publisher      the target publisher
+     * @param instrumenter   the instrumenter
+     * @param classAndMethod the operation name that should be started
+     * @param parentContext  the parent {@link Context}.
+     */
+    public TracingPublisher(Publisher<T> publisher,
+                            Instrumenter<ClassAndMethod, Object> instrumenter,
+                            @Nullable ClassAndMethod classAndMethod,
+                            Context parentContext) {
+        this.publisher = publisher;
+        this.instrumenter = instrumenter;
+        this.classAndMethod = classAndMethod;
+        this.parentContext = parentContext;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> actual) {
+
+        if (instrumenter == null || !instrumenter.shouldStart(parentContext, classAndMethod)) {
+            publisher.subscribe(new Subscriber<T>() {
```

Comment:
Do you need this `Subscriber`? It would need to be `CoreSubscriber` to propagate Reactor context.